feat(doc): document how to build Web UI (#14)
anbox-cloud/cloud-gaming-demo: DEVELOP.md

# Development

This document will guide you through how to
. Set up the local development environment
. Build web UI and configure backend service
. Run the cloud gaming demo for the local development
. Add additional games

## Web UI

> The web UI for the clouding gaming demo is built with [Flutter](https://flutter.dev/).

Author: Fixed.

To set up the development environment and build the UI, make sure that you have Flutter installed on your machine.
Follow the [installation instructions](https://docs.flutter.dev/get-started/install/linux) to install it.

> ### How to build

Reviewer: ```suggestion
### Build the web UI
```
Let's drop the "how to" after all - it would look weird in the other sections.

To build the UI, simply run:

  ```
  cd ui && flutter build web
  ```
This builds the UI and places outputs in the `ui/build/web` directory.
Refer to the official [documentation](https://docs.flutter.dev/get-started/web) for details on how to build web apps with Flutter.

> ### How to deploy

Reviewer: ```suggestion
### Deploy the web UI
```

Author: Fixed.

Create a `static` folder under the `backend/service` directory and copy everything that is in the `ui/build/web` directory into it.

  ```
  mkdir -p backend/service/static
  cp -av ui/build/web/* backend/service/static
  ```

> Those static resources are then served by backend service when you access the web UI.

Reviewer: ```suggestion
The backend service will then serve those static resources when you access the web UI.
```

Author: Fixed

## Backend

> The backend service is not just for serving the web UI, but also interacts with Anbox Stream Gateway and provides a set of restful APIs for clients(web UI) to perform operations like list games or launch a game.

Reviewer: ```suggestion
The backend service does not only serve the web UI, but it also interacts with the Anbox Stream Gateway and provides a set of restful APIs for clients (for example, the web UI) to perform operations like listing games or launching a game.
```

Author: Fixed.

> ### Install the dependency

Reviewer: ```suggestion
### Install the dependencies
```
I understand it's only one, but it sounds much better like this imo.

Reviewer 2: +1

Author: Fixed

> Install the `virtualenv` package to create an isolated python environment that the backend service can run on

Reviewer: ```suggestion
Before you start, install the `virtualenv` package to create an isolated Python environment in which the backend service can run:
```

  ```
  sudo apt-get install python3-pip
  sudo pip3 install virtualenv
  cd backend && virtualenv venv
  ```

### Configure the service

> To access the APIs that are exposed by Anbox Stream Gateway, the backend service needs to know

Reviewer: ```suggestion
To access the APIs that are exposed by the Anbox Stream Gateway, the backend service must know the following information:
```

Reviewer 2: Can we link "Anbox Stream Gateway" and other components to the regular docs?

Author: Done

> . The URL of Anbox Stream Gateway
> . The access token for Anbox Stream Gateway Restful API authentication

Reviewer: ```suggestion
- The URL of the Anbox Stream Gateway
- The access token for the Anbox Stream Gateway Restful API authentication
```

> As Anbox Stream Gateway runs behind traefik and secures the Restful API with HTTPS, its URL can be composed by;

Reviewer: ```suggestion
The Anbox Stream Gateway runs behind traefik and secures the restful API with HTTPS. You can determine its URL by running the following command:
```

  ```
  eval "$(anbox-cloud-appliance internal generate-cloud-info)"

>   gateway_url="https://${CLOUD_PUBLIC_LOCATION}"

Reviewer 2: We should not use the internal `generate-cloud-info` as it is not meant for external use and not considered stable. Instead we can just say `gateway-url="https://<machine public address>"`

Author: +1, Will fix.

  ```
Create an access token to authenticate API requests from a client:

   ```
   gateway_token=$(sudo -u ubuntu juju ssh -m appliance:anbox-cloud anbox-stream-gateway/0 \

>                    -- "sudo /snap/bin/anbox-stream-gateway account create <account-name>")

Reviewer 2: Why don't we use `anbox-cloud-appliance gateway account create <account name>`?

Author: See https://github.com/anbox-cloud/cloud-gaming-demo/blob/main/scripts/install.sh#L76
At the time of writing, we don't release the new image which includes the fix to the marketplace. As we just released the 1.13.1 image today, we can use the above one.
I'll fix the install script as well.

   ```

> Then create a file called `config.yaml` that contains those two config items

Reviewer: ```suggestion
Then create a file called `config.yaml` that contains those two configuration items:
```

   ```
   gateway-url: <gateway_url>
   gateway-token: <gateway_token>
   ```

> Add the following line to the `activate` hook under `backend/venv/bin/` directory so that the backend service can read the `config.yaml` on its startup.

Reviewer: ```suggestion
Add the following line to the `activate` hook in the `backend/venv/bin/` directory so that the backend service can read the `config.yaml` on startup:
```

   ```
   export CONFIG_PATH=<path_to_config_yaml>
   ```

> Then launch the backend service with the following script:

Reviewer: Let's move this to a separate section, because this is how you start the demo, right?

```suggestion

## Run the demo with your changes

To run the demo, launch the backend service with the following script:
```

   ```
   cd backend && ./run.sh
   ```
When the backend service is fully up and running, the web UI will be available at `http://0.0.0.0:8002`.

> ## Adding additional games

Reviewer: ```suggestion
## Add additional games
```

Author: Fixed

> 1. To add an additional game, make a new folder under the `HOME` directory and create a manifest.yaml file in it.

Reviewer: ```suggestion
To add another game to the demo, complete the following steps:

1. Create a folder under the `HOME` directory and create a `manifest.yaml` file in the folder.
```

Author: Fixed

2. Add the following content to the manifest file:

   ```
   name: <game-name>
   instance-type: <instance-type>
   ```

>    This uses the default image as the base to create the application in the AMS.

Reviewer 2: We should add a new line after the code block for consistency. Basically we put a new line before and after each code block

Author: +1, Will fix.

>    This uses the default image as the base to create the application in the AMS.
>    NOTE: please choose the decent [instance-type](https://anbox-cloud.io/docs/ref/instance-types) based on the Anbox Cloud Appliance deployment.

Reviewer: ```suggestion
   This uses the default image as the base to create the application in AMS.
   
   Make sure to choose a suitable [instance type](https://anbox-cloud.io/docs/ref/instance-types) based on the Anbox Cloud Appliance deployment.
```

Author: Fixed.

> 3. Rename the APK file to `app.apk` and put it to the root directory of that folder.
>    And the final layout of the game folder looks as follows:

Reviewer: ```suggestion
3. Rename the APK file to `app.apk` and put it into the root directory of your folder.
   The final layout of the game folder should look as follows:
```

Author: Fixed.

   ```
   .
   ├── app.apk
   └── manifest.yaml
   ```
4. Run the following command from the folder to create the application in AMS:
    ```
    amc application create  .
    ```

>     You can monitor the status of the application creation by

Reviewer: ```suggestion
    You can monitor the status of the application creation with the following command:
```

Author: Fixed

    ```
    amc application ls
    ```

>    After the status of the application is changed to 'ready', the game is created successfully in the AMS.

Reviewer: ```suggestion
   When the status of the application changes to 'ready', the game has successfully been created in AMS.
```

Author: Fixed

> 5. To present the game on the web UI, update the `gameids` variable defined in `ui/lib/homepage.dart` file to include the name of the game declared in the manifest file.

Reviewer: ```suggestion
5. To display the game on the web UI, update the `gameids` variable defined in the `ui/lib/homepage.dart` file to include the name of the game (as declared in the manifest file).
```

Author: Fixed.

   ```
   final gameids = ['bombsquad', 'bbr2', 'mindustry', 'minetest', <game-name>];
   ```

> 6. Insert a new key-value pair to the static `appNameMap` and `appDesMap` variables defined in `lib/api/application.dart` file, respectively representing the display name and description of the game shown on the home page.

Reviewer: ```suggestion
6. Insert a new key/value pair to the static `appNameMap` and `appDesMap` variables defined in the `lib/api/application.dart` file, respectively representing the display name and the description of the game shown on the home page.
```

   ```
   static Map<String, String> appNameMap = const {
   ...
   ...
   '<game-name>': '<display-name>'
   };

   static Map<String, String> appDesMap = const {
   ...
   ...
   '<game-name>': '<description>'
   };

   ```

> 7. Provide a screenshot of the game in the jpeg format, rename it to `<game-name>.jpeg`, and put it into `ui/lib/assets` directory.

Reviewer: ```suggestion
7. Provide a screenshot of the game (in jpeg format), rename it to `<game-name>.jpeg` and put it into the `ui/lib/assets` directory.
```

> 8. Rebuild the web UI, copy the content in the `ui/build/web` folder to `backend/service/static` directory, and refresh the webpage, the game will be available to use.

Reviewer: ```suggestion
8. Rebuild the web UI, copy the contents from the `ui/build/web` folder to the `backend/service/static` directory and refresh the webpage. The game will now be available to use.
```

Author: Fixed

> # Development
> 
> This document will guide you through how to
> 
> . Set up the local development environment
> . Build web UI and configure backend service
> . Run the cloud gaming demo for the local development
> . Add additional games

Reviewer: ```suggestion
# Develop your own demo

This document shows how to customise the demo and add your own changes.
It describes how to build the web UI, configure the backend service, run the demo with your local changes and add additional games.
```
If we want to have a list here, it must correspond fully to the sections below. So imo it's easier to have a sentence describing the content.

Author: I'm fine with that.

> NOTE: this document assumes that you have installed the [Anbox Cloud Appliance](https://anbox-cloud.io/docs/ref/glossary#anbox-cloud-appliance) on your local machine. Follow the [instructions](https://anbox-cloud.io/docs/tut/installing-appliance) on how to install it if you haven't.

Reviewer: ```suggestion
NOTE: This document assumes that you have installed the [Anbox Cloud Appliance](https://anbox-cloud.io/docs/ref/glossary#anbox-cloud-appliance) on your local machine. Follow the [instructions](https://anbox-cloud.io/docs/tut/installing-appliance) on how to install it if you haven't.
```

Author: Fixed.

### Deploy the web UI

Create a `static` folder under the `backend/service` directory and copy everything that is in the `ui/build/web` directory into it.

  ```
  mkdir -p backend/service/static
  cp -av ui/build/web/* backend/service/static
  ```

The backend service will then serve those static resources when you access the web UI.

## Backend

The backend service does not only serve the web UI, but it also interacts with the Anbox Stream Gateway and provides a set of restful APIs for clients (for example, the web UI) to perform operations like listing games or launching a game.

### Install the dependencies

Before you start, install the `virtualenv` package to create an isolated Python environment in which the backend service can run:

  ```
  sudo apt-get install python3-pip
  sudo pip3 install virtualenv
  cd backend && virtualenv venv
  ```

### Configure the service

To access the APIs that are exposed by the Anbox Stream Gateway, the backend service must know the following information:

- The URL of the Anbox Stream Gateway
- The access token for the Anbox Stream Gateway Restful API authentication

> As [Anbox Stream Gateway](https://anbox-cloud.io/docs/ref/glossary#stream-gateway) runs behind traefik and secures the Restful API with HTTPS, its URL can be composed by;

Reviewer: ```suggestion
The [Anbox Stream Gateway](https://anbox-cloud.io/docs/ref/glossary#stream-gateway) runs behind traefik and secures the restful API with HTTPS. Specify its URL in the following way:
```

Reviewer: I don't understand what we're doing here. This is not a command? And we specify the URL below in the config.yaml file?

Author: No, it's not a command. The `gateway_url` here is just linked to the config item(gateway_url) that people need to specify in the config.yaml file. When reading this, it's a bit lengthy and not that clear to me either. Let me improve this or you may have a better idea to describe it in a concise way.

  ```
  gateway_url="https://<machine public address>"
  ```

Create an access token to authenticate API requests from a client:

   ```
   gateway_token=$(anbox-cloud-appliance gateway account create <account-name>)
   ```

Then create a file called `config.yaml` that contains those two configuration items:

   ```
   gateway-url: <gateway_url>
   gateway-token: <gateway_token>
   ```

Add the following line to the `activate` hook in the `backend/venv/bin/` directory so that the backend service can read the `config.yaml` on its startup:

   ```
   export CONFIG_PATH=<path_to_config_yaml>
   ```

## Run the demo with your changes

To run the demo, launch the backend service with the following script:

   ```
   cd backend && ./run.sh
   ```

When the backend service is fully up and running, the web UI will be available at `http://0.0.0.0:8002`.

## Add additional games

To add another game to the demo, complete the following steps:

1. Create a folder under the `HOME` directory and create a `manifest.yaml` file in the folder.

2. Add the following content to the manifest file:

   ```
   name: <game-name>
   instance-type: <instance-type>
   ```

>    This uses the default image as the base to create the application in the [AMS](https://anbox-cloud.io/docs/exp/ams).

Reviewer: ```suggestion
   This uses the default image as the base to create the application in [AMS](https://anbox-cloud.io/docs/exp/ams).
```

>    NOTE: please choose the decent [instance-type](https://anbox-cloud.io/docs/ref/instance-types) based on the Anbox Cloud Appliance deployment.

Reviewer: ```suggestion
   Make sure to choose a suitable [instance type](https://anbox-cloud.io/docs/ref/instance-types) based on the Anbox Cloud Appliance deployment.
```

3. Rename the APK file to `app.apk` and put it into the root directory of your folder.
   The final layout of the game folder should look as follows:

   ```
   .
   ├── app.apk
   └── manifest.yaml
   ```

4. Run the following command from the folder to create the application in AMS:

    ```
    amc application create  .
    ```

    You can monitor the status of the application creation with the following command:

    ```
    amc application ls
    ```

>    When the status of the application changes to 'ready', the game has successfully been created in AMS.

Reviewer: ```suggestion
   When the status of the application changes to "ready", the game has successfully been created in AMS.
```
or
```suggestion
   When the status of the application changes to `ready`, the game has successfully been created in AMS.
```

Author: I vote for the latter one.

> 5. To display the game on the web UI, update the `gameids` variable defined in `ui/lib/homepage.dart` file to include the name of the game (as declared in the manifest file).

Reviewer: ```suggestion
5. To display the game on the web UI, update the `gameids` variable defined in the `ui/lib/homepage.dart` file to include the name of the game (as declared in the manifest file).
```

Author: Fixed.
